feat: Introduce ‘globalStyleMap’

enisdenjo opened the pull request:
### Prelude
`globalStyleMap` is a helper which allows easier creation of multiple "global" styles and theme dependent styles.

### Example 1

Say you have a checkbox component which has custom checkmark icons. Displaying the appropriate checkmark through completely through CSS would be done like so:

```js
// Checkbox.treat.js

import { styleMap, globalStyleMap } from 'treat';

const styles = styleMap({
  checkbox: {
    position: 'relative',
    cursor: 'pointer',
  },
  uncheckedIcon: {
    display: 'flex',
    color: 'inherit',
  },
  checkedIcon: {
    display: 'flex',
    color: 'inherit',
  },
});

globalStyleMap({
  [`${styles.checkbox} > input`]: {
    position: 'absolute',
    cursor: 'pointer',
    opacity: 0,
    margin: 0,
    padding: 0,
  },
  [`${styles.checkbox} > input:checked ~ ${styles.uncheckedIcon}`]: {
    display: 'none',
  },
  [`${styles.checkbox} > input:not(:checked) ~ ${styles.checkedIcon}`]: {
    display: 'none',
  },
});

```

### Example 2

Say you have a theme object looking like this:
```ts
// themed.d.ts

interface Theme {
  palette: {
    primary: string;
    secondary: string;
    accent: string;
  };
}
```
you also have a button which accepts SVGs as direct children. The SVG is `N` amount lighter from the palette background color of the button. You'd approach it like so:

```js
// Button.treat.js

import { styleMap, globalStyleMap } from 'treat';

const styles = styleMap(({ palette }) => ({
  button: {
    display: 'block',
  },
  // maps all colors to button classnames: color-{color}
  // (color-primary, color-secondary, ...etc)
  ...Object.keys(palette).reduce(
    (acc, color) => ({
      ...acc,
      [`color-${color}`]: {
        backgroundColor: palette[color],
      },
    }),
    {},
  ),
}));

globalStyleMap(({ palette }) => ({
  [`${styles.button} > svg`]: {
    display: 'flex',
    fill: 'currentColor',
  },
  // does the same mapping as above, but lightens the
  // direct child SVG color by a factor of 0.5
  ...Object.keys(palette).reduce(
    (acc, color) => ({
      ...acc,
      [`${styles[`color-${color}`]} > svg`]: {
        color: lighten(palette[color], 0.5),
      },
    }),
    {},
  ),
}));
```

mattsjones commented:
Not sure this API makes sense as global styles don't return anything. `styleMap` is named that because it returns a map of styles. 

Also, I feel the current API can implement this scenario just as easily. 

```
for (const color of Object.keys(palette)) {
    globalStyle(`${styles[`color-${color}`]} > svg`, ({ palette }) => ({
        color: lighten(palette[color], 0.5),
    }))
} 
```

enisdenjo commented:
@mattsjones that works for a simple implementation where you have direct access to the theme object. But you actually don't... The theme object is provided through the context you cannot "just" import it.

In your example, where do you get the `palette` for the loop from?

mattsjones commented:
Right, I see what you mean. Had a brain fade in my original example. 

The `globalStyleMap` you're proposing would require every style to feature a themed style, which would be confusing to consumers. `globalStyle` already suffers from this issue but we handle it as `globalStyle` is meant as more an edge-case API. 

Treat encourages you to target a single element with each style. We feel this makes refactoring much easier down the track. In your example you have two elements, the button and the svg. Would the following example be sufficient?

```
const buttonStyles = styleMap(({ palette }) => ({
  button: {
    display: 'block',
  },
  // maps all colors to button classnames: color-{color}
  // (color-primary, color-secondary, ...etc)
  ...Object.keys(palette).reduce(
    (acc, color) => ({
      ...acc,
      [`color-${color}`]: {
        backgroundColor: palette[color],
      },
    }),
    {},
  ),
}));

const svg = style(({ palette }) => ({
  display: 'flex',
  fill: 'currentColor',
  selectors: Object.keys(palette).reduce(
    (acc, color) => ({
      ...acc,
      [`${buttonStyles[`color-${color}`]} > &`]: {
        color: lighten(palette[color], 0.5),
      },
    }),
    {},
  ),
}));
```

enisdenjo commented:
I completely agree with single element targeting for classes and I tend to do so myself. However, in this exact example, I have no control over the `svg` element. It is basically a child of the component which can receive any svg and will style it accordingly.

In order to avoid manipulating the children (which can get messy quite quick), I use plain-ol' CSS to style them (which is the proper, cleaner and more straight-forward way). This is exactly where the `globalStyleMap` comes in handy.

enisdenjo commented:
@mattsjones sorry, I don't want to be annoying. I am contemplating about using this lib for my future styling journeys, because I believe treat is onto something brilliant. However, the missing decision piece is this PR.

It would be a burden for me to fork the lib just to add 25 lines of code. The last thing I want is to go against your conventions, so I fully understand your point; but IMHO I think I am describing something that would elevate a lot of unnecessary restrictions.

mattsjones commented:
Hey bud. Sorry about being slow to reply. I appreciate you being clear in your reasoning why this API makes sense. While I hope this is mostly an edge case API I can see why it makes sense in certain circumstances. 

Only thing is the name. styleMap returns a map like object of styles, hence the name. In this case nothing is returned. I think `globalStylesheet` would suit better. Thoughts?

Also we'll need to add some tests. You just add some calls to the existing test cases and update the snapshots.

markdalgleish commented:
Nitpick, but I'd go with `globalStyleSheet`, similar to how React Native names its `StyleSheet` API.

maraisr commented:
Wonder if this could also be abstracted to solve the #93 issue also 👀

enisdenjo commented:
@mattsjones, @maraisr I have made the name adjustment and added some tests for which I updated the snapshots. I tried keeping the tests as close as possible to the examples I've made here.

markdalgleish commented:
Hey, sorry for the delay on this.

Had a dig into this and I think there's a couple of things that still need to be addressed.

**First, we need to ensure that all selectors are either themed or unthemed—you can't have a mix.**

For example, this *shouldn't* be valid:

```js
export const themedClass = style(theme => { ... });

globalStyleSheet(theme => ({
  [`${themedClass} svg`]: { ... },
  '.nonThemedClass': { ... }
});
```

Whereas this *should* be valid:

```js
export const themedClass = style(theme => { ... });

globalStyleSheet(theme => ({
  [`${themedClass} svg`]: { ... },
  [`${themedClass} .nonThemedClass`]: { ... }
});
```

In `globalStyle` we throw an error when you try to use themed styles without a themed selector if you'd like to reference existing patterns for this.

**Secondly, we need to ensure that you can't use themed styles _within_ a global style sheet object, both logically and from a TypeScript perspective.**

For example, this should be a type error :

```js
globalStyleSheet(theme => ({
  [`${themedClass} svg`]: theme => ({ ... })
});
```

For consistency, this should also be a type error:

```js
globalStyleSheet({
  [`${someClass} svg`]: theme => ({ ... })
});
```

Note that we obviously have non-TypeScript consumers, so we'd also need to throw a runtime error in both of these cases.

Hope that makes sense. This API is surprisingly tricky 😬

enisdenjo commented:
Makes perfect sense! No worries, Ill add tests and implement the two points.

Sadly, I am clogged with work currently, so I cannot jump on it rather soon; but know that I am happy to help out.
